### ISSUE TITLE
API Changes

### ISSUE THREAD

**Issue by kingosticks** [mopidy/mopidy-soundcloud#123]
Soundcloud devs seems to be alive! They've made / are making some API changes which look like they will impact us. The API docs have moved [here](https://developers.soundcloud.com/docs/api/explorer/open-api). The old endpoints still seem to work as of now. They also seem to be doing something to [pagination](https://developers.soundcloud.com/blog/pagination-updates-on-our-api) and it sounds like the default response format is going to change.

**Comment by djmattyg007** [comment]
How does this affect the PRs that @Laurentww has submitted recently?

**Comment by kingosticks** [comment]
I've only looked at the rate limit one. Only the image one seems to hit the official API and that's restricted to the tracks endpoint which I don't think changed. Looks like it's mostly the /me/foo endpoints.

**Comment by Laurentww** [comment]
Adding images (#121) still uses the official API, as the standard app client id doesn't return rate errors here. Though, #121 could be improved as well by first attempting with the publicly obtained client id and reversing to the standard app client id on failure.

As far as I can see, no pagination calls are made and the current API call uris are up to date.

**Comment by kingosticks** [comment]
More api changes coming https://developers.soundcloud.com/blog/security-updates-api

**Comment by bretonium** [comment]
According to soundcloud/api#151, this will not be fixed. Soundcloud api can now be used only together with a service that will issue a token. The tokens also need to be refreshed, since they are now time-limited.

Can mopidy provide an API endpoint that the clients could use to fetch the tokens? Is there anything similar in mopidy already?

**Comment by kingosticks** [comment]
Yes, Mopidy-Spotify works this way (utilising https://github.com/adamcik/oauthclientbridge hosted at auth.mopidy.com). We should be able to re-use that here also. Although I am not sure who controls Mopidy's soundcloud app and if they would need to update anything in the account settings, maybe it is @adamcik ?

Although they (Soundcloud) maybe need to sort out their CORS issues also (from https://github.com/soundcloud/api/issues/151).

**Comment by tkem** [comment]
@adamcik Well, sorry, no access.

**Comment by kingosticks** [comment]
@dz0ny has access and can provide he secret (is that all we need @adamcik) but everything else is locked out and can't be changed. This includes the callback URL but hopefully that isn't a show-stopper for us.

**Comment by adamcik** [comment]
That should be workable with the id + secret

**Comment by adamcik** [comment]
Thanks @kingosticks for getting the secret to me, I tried running this in my local test instance of the oauth bridge and I need to do one minor code change and it should all just work.

Only ugly bit we'll have to do is to update https://mopidy.com/soundcloud_callback with some JS and a fallback link that sends people to https://auth.mopidy.com/soundcloud/callback with the same parameters and then we are back in business.

So what we need to progress is:

- [ ] List of the scopes we should ask for
- [ ] Setting up the oauthclientbridge properly
- [ ] Update of soundcloud_callback to redirect if the "right" oauth param is there so we can handle old and new cases
- [ ] Update of the plugin to start using the new flow with refresh tokens
- [ ] Update of the extension page on mopidy.com to fetch the new token types

What did I miss?

**Comment by adamcik** [comment]
And by should just work I mean that I did manage to get a working token in my locally modified test setup.

**Comment by chris-aeviator** [comment]
is there a way where users can setup their own bridge in order to get soundcloud running again?

**Comment by adamcik** [comment]
Sorry for the delay, I haven't had much chance to work on my part of this.

@chris-aeviator no not really, my code is on github, but you need credentials from soundcloud that they've stopped handing out years ago :(

**Comment by chris-aeviator** [comment]
Ok - thanks for clarification - I can see that I can't register a new API account...bummer that soundcloud forces us to use their crappy player

**Comment by adamcik** [comment]
https://auth.mopidy.com/soundcloud is now up and running - since we can't change the redirect it sends us to https://mopidy.com/soundcloud_callback/?code=REDACTED&state=REDACTED and not https://auth.mopidy.com/soundcloud/callback as it's supposed to. I.e. next step is to add JS to `/soundcloud_callback` to detect that we are in the wrong place and send people to the right domain so the rest of the flow can complete.

With the bridge up and running none of the remaining work should be blocked on me, other than the one minor code change to the bridge that I only have in my dev setup. If we need to specify a default scope just let me know, and I'll update the config. But at least other contributors can pickup the other items I mentioned further up.

**Comment by adamcik** [comment]
The old case sends you to https://mopidy.com/soundcloud_callback/?code=REDACTED&state=REDACTED#access_token=REDACTED&scope=non-expiring and fails with eg https://mopidy.com/soundcloud_callback/?error_code=access_denied&error_description=user%20cancelled%20request&state=REDACTED

So for successful requests we can check if the hash is set and decide if we should redirect or not, but for errors I think our only option is to encode something in our state parameter :/

**Comment by adamcik** [comment]
Testing with changing the button href in https://mopidy.com/ext/soundcloud/ to point at the bridge and going through ends up failing with `INVALID_STATE: Authentication page sent wrong state to pop-up.` since we get the state that the bridge and soundcloud share, not the mopidy website and the bridge.

So perhaps we can just blindly assume that we should redirect this case to the bridge, and worst case it also fails the state check and no harm done, and in most cases we end up with a page that can handle the old and new cases.

**Comment by nimbling** [comment]
Hopefully this helps:
I had the `mopidy_soundcloud.soundcloud Invalid "auth_token" used for SoundCloud authentication!` line in my Mopidy log, but that disappeared after I authorized myself at [https://developers.soundcloud.com/docs/api/explorer/open-api](https://developers.soundcloud.com/docs/api/explorer/open-api)

**Comment by simaovienna** [comment]
As soundcloud "...will no longer be processing API application requests at this time...":

https://soundcloud.com/you/apps/new

is there any chance/workaround to get the integration to work? @nimbling how did you authorize?

thanks

**Comment by nimbling** [comment]
I didn't, please see my correction. I did just log in using that link here: https://developers.soundcloud.com/docs/api/explorer/open-api - but it did not help as much as I thought it did.

**Comment by zlocate** [comment]
After removing the client_id from the request url in the API, everything worked.

**Comment by Nick1296** [comment]
> After removing the client_id from the request url in the API, everything worked.

Same here! 
I just removed the client_id in `steamble_url` and `_get`

**Comment by kingosticks** [comment]
And are they still working? According to the SC docs those tokens will expire after an hour. If that is true then you'd have to go and re-auth Mopidy to get another one.

**Comment by Nick1296** [comment]
Tried loading a random song right now from soundcloud, it seems to be working.
I'll keep trying here and there to see when the token expires

**Comment by kingosticks** [comment]
If people wouldn't mind giving https://github.com/mopidy/mopidy-soundcloud/pull/133 some test time that'd be handy. I don't really use it.

**Comment by photoangell** [comment]
Pulled your fix and it works great for me. 
I did get a one time log entry when first playing a song...
```
INFO     2022-05-20 13:47:29,697 [7:SoundCloudBackend-18] mopidy_soundcloud.soundcloud
  Streaming of public song using public client id failed, trying with standard app client id..
```
but it doesn't seem to affect playback and I haven't seen the message since.
my config is
```
[soundcloud]
enabled = true
auth_token = xxxxxxxxxxxxxxxxxxxxxxxxxx
explore_songs = 100
```

**Comment by kingosticks** [comment]
Yep, that message is normal. And it shows you are actually using the (standard) logged in client ID, which is the one we have hopefully now fixed.

**Comment by BlackLight** [comment]
How did you folks manage to get an `auth_token`? After authenticating to https://mopidy.com/ext/soundcloud/ I don't get any usable token (nor errors on the console). I have tried with the `oauth_token` from an active web session too, but that didn't work either.

**Comment by Homer-Sim** [comment]
Sorry for my question, how do I pull a fix?

**Comment by kingosticks** [comment]
Search online how to install from git using pip. Failing that I'll cut a release tonight. I have my doubts about this hack and we probably still need to do the real fix that @adamcik laid the groundwork for.

@BlackLight it works just the same as our Spotify auth. Click the button, do what the Soundcloud login popup says, that popup will then close and an auth token have appeared in the box for you to put in your config. I think the page says all this, not sure what could be going wrong as I just checked and it's working fine.

**Comment by Homer-Sim** [comment]
> Search online how to install from git using pip. Failing that I'll cut a release tonight. I have my doubts about this hack and we probably still need to do the real fix that @adamcik laid the groundwork for.

I did a apt remove of existing installation and a 
pip3 install "git+https://github.com/mopidy/mopidy-soundcloud.git"
which tells me about a successful installation
service restarted 
but no soundcloud options in webgui?

Did I something wrong?

**Comment by kingosticks** [comment]
As you're using the service you need to do `sudo pip3 blah blah` for a system-wide installation else the system service user cannot see the installation. You can just redo your command with the added prefix. I'm sorry, I should have mentioned that as it's not obvious and it's particular to programs running as other users (e.g. services).

**Comment by johnhamelink** [comment]
I'm excited to get Soundcloud working with my mopidy setup!

I've tried creating a new auth token for myself, but after I complete login and the popup closes, I get an alert box with the following:

> `INVALID_STATE: Authentication page sent wrong state to pop-up.`

**Comment by dominikdeka** [comment]
@kingosticks 
I have installed Mopidy-Spotify from master:
`sudo python3 -m pip install "git+https://github.com/mopidy/mopidy-soundcloud.git"`
I have restarted Mopidy service for many times, I have generated a new auth_token... and I am still getting `[SoundCloudBackend-6] mopidy_soundcloud.soundcloud Invalid "auth_token" used for SoundCloud authentication!`

Am I missing something?

**Comment by dominikdeka** [comment]
I have installed it again this time with command:
`sudo python3 -m pip install "git+https://github.com/mopidy/mopidy-soundcloud@master"` and this finally replaced the lib with master version. Its great to see mopidy-soundcloud working again. Especially considering the fact that workaround for mopidy-spotify is still in proress. 
Thank you @kingosticks !

**Comment by kingosticks** [comment]
Sounds like we need a release.

**Comment by phijor** [comment]
Is there a chance we can get a new patch release that includes the API changes? I maintain the `mopidy-soundcloud` package in the [AUR](https://aur.archlinux.org/packages/mopidy-soundcloud), and I would prefer not having to manually patch that package.

If there's any blockers that I can help resolve, please tell me :)

Thanks to everyone involved in making this extension!

**Comment by WhitteGoose** [comment]
_Hello all, sorry if I ask stupid questions, I'm beginner in GNU/Linux and wanna listen soundcloud with mopidy._

so, **which token should I put into "auth_token"?** 
`1-1111111-111111111`, which is on readme.md of this repo, or `1-11111-11111111111-abcdvteewq323176dfd `which is on the [mopidy-soundcloud auth page](https://mopidy.com/ext/soundcloud)

offtopic question: how do I turn off mopidy.service and use mopidy as user?

**Comment by dominikdeka** [comment]
In order to obtain authenticate key go to https://mopidy.com/ext/soundcloud/ and utilize `Authenticate mopidy with SoundCloud` button.

> and did I understand correctly that this is the right way to install mopidy-soundcloud now: `sudo python3 -m pip install "git+https://github.com/mopidy/mopidy-soundcloud@master"`

Yeap, it worked for me two days ago. (I am afraid new realese still not yet ready. :( )

> _offtopic question: how do I turn off mopidy.service and use mopidy as user?_ _do I just `systemctl disable mopidy` or what?_

Yes, but as a root `sudo systemctl disable mopidy`. Also make sure you stop the service, before launching mopidy as user: `sudo systemctl stop mopidy`

**Comment by goosegit11** [comment]
Mopidy says that auth key is invalid

**Comment by goosegit11** [comment]
@dominikdeka 

Mopidy says that auth key is invalid

**Comment by goosegit11** [comment]
Jesus Christ, seems like I got it working now :skull: 

### For those who haven't figured that shit out yet:
don't use mopidy-soundcloud from AUR,
don't use just `sudo pip install mopidy-soundcloud`, it's not working.

### Use this:
`sudo python3 -m pip install "git+https://github.com/mopidy/mopidy-soundcloud@master"`

**Comment by kingosticks** [comment]
So... I never made that release?! Yikes. Sorry

**Comment by dreamlayers** [comment]
There is no need to run pip as root via sudo. It only needs to be accessible to the user running mopidy. I used `sudo -u mopidy bash` to become the mopidy user and then installed without sudo, with the install ending up under `~mopidy/.local/lib/`.

**Comment by kingosticks** [comment]
I don't know what to do about this extension going forward. I can do this long overdue release tonight (I hope) but it's still fundamentally crippled by their incredibly stupid rate limiting. They (Soundcloud) keep banging on about how they are working away on the API but they re-iterated just last week they've no plans to open it up again. I don't get the point.

**Comment by herrernst** [comment]
A new release would be nice, thanks in advance.

**Comment by kingosticks** [comment]
https://github.com/mopidy/mopidy-soundcloud/releases/tag/v3.0.2

**Comment by herrernst** [comment]
Thank you very much! Do you know if it will appear automatically in the apt repos at apt.mopidy.com? Here I see only 3.0.1: https://apt.mopidy.com/pool/

**Comment by kingosticks** [comment]
No, it's not automatic. It's a manual extra step, will request for it to happen. Should be in PyPI already though.

**Comment by dominikdeka** [comment]
Thank you @kingosticks for preparing this release!

**Comment by BlackLight** [comment]
Is anything broken again with this integration?

It worked for a few days but it's failing again now, no matter the track I'm trying to play.

I have tried both the latest pip version and the git version just in case. I can log in to SC with my token, browse my music collections and all, but whenever I try to play a track I get this:

```
DEBUG    2023-03-24 10:26:26,524 [798355:SoundCloudBackend-10 (_actor_loop)] mopidy_soundcloud.soundcloud
  Requested https://api.soundcloud.com/tracks/256913562
DEBUG    2023-03-24 10:26:26,787 [798355:SoundCloudBackend-10 (_actor_loop)] mopidy.backend
  Backend translated URI from soundcloud:song/We Have Lost The Way (Instrumental).256913562 to None
WARNING  2023-03-24 10:26:26,787 [798355:Core-12 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/We Have Lost The Way (Instrumental).256913562
```

That `Backend translated URI from soundcloud:song/... to None` part is very suspicious. It seems like something goes very wrong when mapping the SC item to a mopidy item.

Before I start diving into this can of worms, has anybody been experiencing the same issue?

**Comment by kingosticks** [comment]
Working for me. Where is the output from https://github.com/mopidy/mopidy-soundcloud/blob/master/mopidy_soundcloud/soundcloud.py#L245 in your debug log?

**Comment by kingosticks** [comment]
I think that's a normal (or at least functioning) part of the public stream client code. I get:

```
Mar 24 13:05:57 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy_soundcloud.soundcloud Parsing track soundcloud:song/Disappear.284103524
Mar 24 13:05:57 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy_soundcloud.soundcloud Getting info for track with ID 284103524
Mar 24 13:05:57 raspberrypi mopidy[30917]: DEBUG    [MainThread] mopidy.audio.gst Got STATE_CHANGED bus message: old=GST_STATE_NULL new=GST_STATE_READY pending=GST_STATE_VOID_PENDING
Mar 24 13:05:58 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://api.soundcloud.com:443 "GET /tracks/284103524 HTTP/1.1" 200 1039
Mar 24 13:05:58 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy_soundcloud.soundcloud Requested https://api.soundcloud.com/tracks/284103524
Mar 24 13:05:58 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool Starting new HTTPS connection (1): soundcloud.com:443
Mar 24 13:05:58 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://soundcloud.com:443 "GET / HTTP/1.1" 200 None
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool Starting new HTTPS connection (1): a-v2.sndcdn.com:443
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/53-aed4402d.js HTTP/1.1" 200 None
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/51-f3c4bf71.js HTTP/1.1" 200 None
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool Starting new HTTPS connection (1): cdn.cookielaw.org:443
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://cdn.cookielaw.org:443 "GET /consent/7e62c772-c97a-4d95-8d0a-f99bbeadcf61/otSDKStub.js HTTP/1.1" 200 7684
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/16-c0f3e8c3.js HTTP/1.1" 200 None
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/52-dacb4af2.js HTTP/1.1" 200 None
Mar 24 13:05:59 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/1-ac706f12.js HTTP/1.1" 200 None
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/3-0b89809f.js HTTP/1.1" 200 None
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy_soundcloud.soundcloud Updated SoundCloud public client id to: RMDIzNoU4QIzQsT3xq9J5TdxFFQlJvLY
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://soundcloud.com:443 "GET /young-legionnaire/disappear-1 HTTP/1.1" 200 None
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool Starting new HTTPS connection (1): api-v2.soundcloud.com:443
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://api-v2.soundcloud.com:443 "GET /media/soundcloud:tracks:284103524/6e05c0df-dac1-47c5-900c-37e8a738480b/stream/progressive?client_id=RMDIzNoU4QIzQsT3xq9J5TdxFFQlJvLY HTTP/1.1" 403 28
Mar 24 13:06:00 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://soundcloud.com:443 "GET / HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/53-aed4402d.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/51-f3c4bf71.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://cdn.cookielaw.org:443 "GET /consent/7e62c772-c97a-4d95-8d0a-f99bbeadcf61/otSDKStub.js HTTP/1.1" 200 7684
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/16-c0f3e8c3.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/52-dacb4af2.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/1-ac706f12.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://a-v2.sndcdn.com:443 "GET /assets/3-0b89809f.js HTTP/1.1" 200 None
Mar 24 13:06:01 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy_soundcloud.soundcloud Updated SoundCloud public client id to: RMDIzNoU4QIzQsT3xq9J5TdxFFQlJvLY
Mar 24 13:06:02 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] urllib3.connectionpool https://api-v2.soundcloud.com:443 "GET /media/soundcloud:tracks:284103524/6e05c0df-dac1-47c5-900c-37e8a738480b/stream/progressive?client_id=RMDIzNoU4QIzQsT3xq9J5TdxFFQlJvLY HTTP/1.1" 200 547
Mar 24 13:06:02 raspberrypi mopidy[30917]: DEBUG    [SoundCloudBackend-4] mopidy.backend Backend translated URI from soundcloud:song/Disappear.284103524 to https://cf-media.sndcdn.com/FdsWYJXWHnVP.128.mp3?Policy=eyJTdGF0ZW1lbnQiOlt7IlJlc291cmNlIjoiKjovL2NmLW1lZGlhLnNuZGNkbi5jb20vRmRzV1lKWFdIblZQLjEyOC5tcDMqIiwiQ29uZGl0aW9uIjp7IkRhdGVMZXNzVGhhbiI6eyJBV1M6RXBvY2hUaW1lIjoxNjc5NjYzNDg1fX19XX0_&Signature=VysjZz98E8-7--aJYOj2zUC5kCi544NKKCJB~ToapCj3WOBvLgefWS1037g4yx55XL~NycT4cuRhOAeDxHK9Mge6-3pdZws-aUHne2esNA4f~C2H5LFwcG75rn9JXdKrZ1yLri77jLaK~MT3nua7RVBrpRFCNGdWRK-npe84qVUBxAUqxsDyEIutx9-PH~MufHWu0LWRYP8HmlNS5efyYaFCwSqeMHxpJHTyYqSzteKaAe3CFEVEPGbP6WiMBeJljlwtv1-yOo2ds4SeVR~oLGZD33tjckEE3zOJoF8q0IfW1HhEtHr8lsIXJ8QlnunMVGXiWCJZD6Sycabo-~59wQ__&Key-Pair-Id=APKAI6TU7MMXM5DG6EPQ
```

**Comment by BlackLight** [comment]
Hmm I think I know why.

Your log traces look the same as mine for the first part, but I don't get beyond the `GET` to `cdn.cookielaw.org` part.

And that's most likely because I use PiHole on my network, which redirects all the requests to that domain to localhost (I just don't like being harassed by cookie banners).

So it seems like the original request to SoundCloud at some point triggers a request to cookielaw.org, which fails, and therefore the whole parsing logic fails.

First of all, it'd be interesting to know why a backend request gets a redirect to cookielaw.org - as my voice assistant would say, "I can't display you a banner, I don't have a screen!".

Secondly, maybe it's worth skipping those connections on the extension's side as well, if at some point we get that redirect?

A workaround for now can be to whitelist `cdn.cookielaw.org` on my DNS and block it on the frontend side via NoScript instead, but it's probably worth skipping that logic entirely on the backend (or at least not make the upstream request fail if that particular request fails), since it's not functional to the purpose of the operation.

**Comment by kingosticks** [comment]
Maybe, but only until they change cdn.cookielaw.org to cdn1.cookielaw.org. And what if there was some external JS that was required (maybe updating the page contents to provide the data we are trying to extract)? I'm not saying these are not good ideas, I am just pointing out it's all a bit brittle.

I think the first thing would be to make the fallback more reliable. In your case, I assume it gets an empty `200` response for that page and then should just carry on as usual. I don't have time to look into this but I will review and merge a fix.

**Comment by BlackLight** [comment]
> In your case, I assume it gets an empty 200 response for that page and then should just carry on as usual.

It's more like a `404` in my case - because the domain redirects to localhost and I have a web server running on it. Otherwise, it would just be a timed out request. Maybe it's worth checking if any of the nested requests fail, and if those are associated to static files (like images and `.js`) perhaps the logic should just move on?

**Comment by kingosticks** [comment]
Ahh ok so it's trying to update the public client_id that's needed when making requests to https://api-v2.soundcloud.com/foo for foo streams it just scraped from the public page (i.e,. not from API calls). 

In my case it finds the client_id in `<script crossorigin src="[https://a-v2.sndcdn.com/assets/3-0b89809f.js](view-source:https://a-v2.sndcdn.com/assets/3-0b89809f.js)"></script>`. In your case the blocked request must be causing an exception which is silently caught by https://github.com/mopidy/mopidy-soundcloud/blob/c61060abd567e2ddf9ee4591b93bd0a20804a9fc/mopidy_soundcloud/soundcloud.py#L238-L241. 

`except Exception:` is never a good idea.

**Comment by kingosticks** [comment]
So, We should
a) catch the exceptions properly and generally handle errors better in this code
b) possibly ignore cdn.cookielaw.org? It might stop being useful but I can't imagine it'll ever break anything,

**Comment by BlackLight** [comment]
I completely agree on (at the very least) log the exception instead of silently swallowing it - that would have definitely saved me some debugging time.

It sounds to me like some of the JS is actually required, so my initial idea of just skipping `<script>` URLs may not work.

Skipping `cdn.cookielaw.org` sounds like a sensible option, but I agree with you that it may also be a brittle one - all it takes is a domain name to change to throw a wrench in the logic.

If the extension only needs to scrape the page's static JS because it needs a particular one (the `sndcdn.com` link that contains the `client_id`) then maybe it can also make sense to just look for a `<script>` tag that matches the criteria, instead of downloading all of them?

**Comment by kingosticks** [comment]
We could only download scripts from sndcdn.com (there are a bunch), but again, that domain could change and then it would be broken. Unfortunately, other than this cookie one, the scripts mostly look the same and if we decide on some arbitrary criteria that works today it could easily be wrong tomorrow. I think as long as we handle (and log) errored requests we'll be in much better shape. I'll hack a fix in and make an issue to fix it properly.  Anyone else that hits this is very welcome to take it on.

**Comment by Goosegit11** [comment]
Search is working for me, but playing tracks are not... Running these versions (I think they are the latest):

```
Mopidy                          3.4.1
Mopidy-MPD                      3.3.0
Mopidy-SoundCloud               3.0.2
```

Log:
```
INFO     2023-12-09 03:05:00,858 [1342184:MpdSession-14 (_actor_loop)] mopidy_mpd.session
  New MPD connection from [::ffff:127.0.0.1]:46502
ERROR    2023-12-09 03:05:09,424 [1342184:Core-7 (_actor_loop)] mopidy.core.library
  FileBackend backend returned bad data: Expected a SearchResult instance, not []
ERROR    2023-12-09 03:05:09,424 [1342184:Core-7 (_actor_loop)] mopidy.core.library
  StreamBackend backend returned bad data: Expected a SearchResult instance, not []
INFO     2023-12-09 03:05:09,425 [1342184:SoundCloudBackend-6 (_actor_loop)] mopidy_soundcloud.library
  Searching SoundCloud for: серега
WARNING  2023-12-09 03:05:23,179 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:30,916 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:31,910 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:32,382 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:32,836 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:33,250 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467
WARNING  2023-12-09 03:05:33,656 [1342184:Core-7 (_actor_loop)] mopidy.core.tracklist
  Track is not playable: soundcloud:song/-.1131515467

```
